Js with class

Vagyasri opened the pull request:
**Manage books collection (plain JS with classes):**
  > In this project, we refactored the Awesome books application to use JavaScript classes.
  > In this process we learned the following
-   how to use them, their syntax
-   relationship
-   differences with objects and functions

Vagyasri committed:
Update Readme

elasfarc committed:
factoring-classes

Vagyasri committed:
Create Book Class and instances

Vagyasri committed:
Fix linter errors

elasfarc committed:
Add Book class

elasfarc committed:
Add #add in Methods

elasfarc committed:
Add #display on methods class

elasfarc committed:
#handleAddtion (form submission listener  )

Vagyasri committed:
Add remove method

Vagyasri committed:
Fix linter

Vagyasri committed:
Update Readme

hemantso commented on index.js:
```diff
@@ -1,57 +1,65 @@
+    }
+  }
+
+  static handleAddtion(event) {
+    event.preventDefault();
+    const titleInput = document.querySelector('input[placeholder="Title"]');
+    const authorInput = document.querySelector('input[placeholder="Author"]');
+    if (!(titleInput.value.length < 3 || authorInput.value.length < 3)) {
+      Methods.add({ title: titleInput.value, author: authorInput.value });
+      location.reload();
+    } else {
+      const form = document.querySelector('form');
+      const errorDiv = document.createElement('div');
+      form.prepend(errorDiv);
+      errorDiv.innerHTML = `
```
- [x] Please try to avoid manipulating the DOM inside the class. Let's try to fix it by creating a function outside the class and manipulating the DOM for the required result in this function.

hemantso commented on index.js:
```diff
@@ -1,57 +1,65 @@
-    const errorDiv = document.createElement('div');
-    form.prepend(errorDiv);
-    errorDiv.innerHTML = `
+  static display() {
+    const booksContainer = document.getElementById('display-books');
+    // eslint-disable-next-line no-restricted-syntax
+    for (const book of Book.listOfBooks) {
+      booksContainer.innerHTML += `
+          <div id='${book.serialNumber}'>
+              <p>${book.title}</p>
+              <p>${book.author}</p>
+              <button type="button" class="removebtn" onclick='Methods.remove(event)'> Remove </button>
+              <hr>
+          </div> 
+          `;
+    }
+  }
```
- [x] Please try to avoid using the DOM manipulation outside the Class. It will help us to make the code cleaner.

hemantso commented on index.js:
```diff
@@ -1,57 +1,65 @@
-  listOfBooks = JSON.parse(localStorage.getItem('books'));
-  // eslint-disable-next-line eqeqeq
-  listOfBooks = listOfBooks.filter((book) => book.serialNumber != remSerial);
+  static add({ title, author }) {
+    // eslint-disable-next-line no-new
+    new Book({ title, author });
+    localStorage.setItem('books', JSON.stringify(Book.listOfBooks));
+  }
```
- [x] Static method is directly called on the class. Static methods are that using them forfeits one of the advantages of objects. Objects are intended for data encapsulation. Instead of using the static method `add`, we can attach an array property with class and push parameters of the `add` method inside this property. And we return the array or we can get the updated array from the getter method.

hemantso commented on index.js:
```diff
@@ -1,57 +1,65 @@
 
-  localStorage.setItem('books', JSON.stringify(listOfBooks)); location.reload();
-}
+  static remove(event) {
+    const remSerial = event.target.parentElement.id;
+    // eslint-disable-next-line eqeqeq
+    Book.listOfBooks = Book.listOfBooks.filter((book) => book.serialNumber != remSerial);
```
- [x] Kindly don't use the static method. Static methods are that using them forfeits one of the advantages of objects. Objects are intended for data encapsulation. We can use the array inside the class as property and then adding and deleting the book from the array much easier. we can get an updated array as a return from the `add` and `remove` method or we can use the getter method.

hemantso reviewed (changes requested):
Hi @Vagyasri and @elasfarc, 

Good job so far! 👍 
There are some issues that you still need to work on to go to the next project but you are almost there! 

![Alt Text](https://media.giphy.com/media/q6U4fbreC4Neo/giphy.gif)
## Required Changes ♻️
_Check the comments under the review._

- [x]  Use DOM manipulation outside the class.
- [x] Changes in `add` method
- [x] Changes in `remove ` method

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

You can also consider:
- Use array methods that don't mutate the array, e.g. concat, filter, slice instead of push, pop, splice and so on. 

Cheers and happy coding 😃

Feel free to leave any questions or comments in the PR thread if something is not 100% clear. You can mention me (@hemant-soni-vst-au4) so I can see your comments and respond.


**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

elasfarc committed:
separate DOM manipulating from class implementation

-integrate #add, #remove class Methods into Book class ( the Book Class is responsible to add to and/or remove from its book collection)

mimipeshy reviewed (changes requested):
## Hi @Vagyasri and @elasfarc 👋🏾 👋🏾  , 

Great work on making the changes requested by a previous reviewer 👏 

You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.

 ## Required Changes ♻️

- [x] I have noticed you have two js files. One uses static which the previous TSE discouraged. The other uses objects which is the recommended option for this milestone. Kindly do away with the one that uses classes and adopt the one that uses objects.

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._
```

elasfarc committed:
separate function the books into new class Library handles Book objs

elasfarc committed:
Add UI helper methods to handle events and display

elasfarc committed:
fix linter errors

elasfarc committed:
Modify assigning new book's ID

carlos-ssh commented on .vscode/settings.json:
```diff
@@ -0,0 +1,3 @@
+{
+    "liveServer.settings.port": 5501
+}
```
[OPTIONAL]
- Add directories that aren't from your project to your .gitignore file. (.vscode)

carlos-ssh reviewed (approved):
Hi @Vagyasri & @elasfarc, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 



Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have only 3 code reviews per project. If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

SpaYco commented:
Hello 👋,

This is an additional review as a result of the second opinion request.

On this project, you used static methods for your class, which makes them able to call the function directly from their class without making `const methods = new classNameExample` which isn't the way it's supposed to be done here, 

However, the review messages were a bit confusing, sorry about that!
I would suggest you go back to the code and correct the classes you get the experience, this is optional as you already got the project approved, but I highly recommend doing it!

All the best, Keep rocking 🚀!
